src/basic/fs-util.c

```diff
@@ -967,3 +967,33 @@ int unlinkat_deallocate(int fd, const char *name, int flags) {
 
         return 0;
 }
+
+int fsync_directory_of_file(int fd) {
+        _cleanup_free_ char *path = NULL, *dn = NULL;
+        _cleanup_close_ int dfd = -1;
+        int r;
+
+        r = fd_verify_regular(fd);
+        if (r < 0)
+                return r;
+
+        r = fd_get_path(fd, &path);
+        if (r < 0)
+                return r;
+
+        if (!path_is_absolute(path))
+                return -EINVAL;
+
+        dn = dirname_malloc(path);
```

Comment:
What entity calls `free()` on `dn`?

Reply:
@portante note the _cleanup_free_ qualifier on dn's type and how it's initialized to NULL.  This is magic that frees dn when the function returns if dn is != NULL.